File: scripts/day1/day1_scripts_Python/Day1_Application3_NIH.py
import os
import json
import pandas as pd
import requests

# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 1 - From web source to raw dataset
# Application 3: Life Sciences - NIH RePORTER API
# =============================================================================
# Goal: Extract raw structured data about funded research grants.

def fetch_nih_data():
    """
    Fetches NIH RePORTER data for projects related to 'genomics'.
    Saves the raw JSON response to a local cache file.
    """
    cache_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data', 'cache')
    os.makedirs(cache_dir, exist_ok=True)
    cache_file = os.path.join(cache_dir, 'day1_app3_nih_raw.json')

    if os.path.exists(cache_file):
        print(f"Loading cached data from {cache_file}")
        with open(cache_file, 'r', encoding='utf-8') as f:
            raw_data = json.load(f)
    else:
        print("Fetching fresh data from NIH RePORTER API...")
        url = "https://api.reporter.nih.gov/v2/projects/search"
        payload = {
            "criteria": {
                "advanced_text_search": {
                    "operator": "and",
                    "search_field": "terms",
                    "search_text": "genomics"
                }
            },
            "limit": 50,
            "offset": 0
        }
        headers = {"Content-Type": "application/json"}
        
        response = requests.post(url, json=payload, headers=headers)
        response.raise_for_status()
        raw_data = response.json()
        
        with open(cache_file, 'w', encoding='utf-8') as f:
            json.dump(raw_data, f, indent=4)
        print(f"Saved raw data to {cache_file}")

    if 'results' in raw_data and len(raw_data['results']) > 0:
        first_record = raw_data['results'][0]
        print("\nExample of raw extracted record:")
        print(json.dumps(first_record, indent=2))
    else:
        print("No records found.")

if __name__ == "__main__":
    fetch_nih_data()

File: scripts/day1/day1_scripts_R/Day1_Application3_NIH.R
# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 1 - From web source to raw dataset
# Application 3: Life Sciences - NIH RePORTER API
# =============================================================================
# Goal: Extract raw structured data about funded research grants.

library(httr2)
library(jsonlite)
library(fs)

fetch_nih_data <- function() {
  cache_dir <- path(path_dir(path_dir(getwd())), "data", "cache")
  dir_create(cache_dir)
  cache_file <- path(cache_dir, "day1_app3_nih_raw.json")
  
  if (file_exists(cache_file)) {
    cat("Loading cached data from", cache_file, "\n")
    raw_data <- fromJSON(cache_file, simplifyVector = FALSE)
  } else {
    cat("Fetching fresh data from NIH RePORTER API...\n")
    url <- "https://api.reporter.nih.gov/v2/projects/search"
    
    payload <- list(
      criteria = list(
        advanced_text_search = list(
          operator = "and",
          search_field = "terms",
          search_text = "genomics"
        )
      ),
      limit = 50,
      offset = 0
    )
    
    req <- request(url) %>%
      req_headers("Content-Type" = "application/json") %>%
      req_body_json(payload)
    
    resp <- req_perform(req)
    raw_data <- resp_body_json(resp, simplifyVector = FALSE)
    
    write_json(raw_data, cache_file, pretty = TRUE, auto_unbox = TRUE)
    cat("Saved raw data to", cache_file, "\n")
  }
  
  if (!is.null(raw_data$results) && length(raw_data$results) > 0) {
    cat("\nExample of raw extracted record:\n")
    print(toJSON(raw_data$results[[1]], pretty = TRUE, auto_unbox = TRUE))
  } else {
    cat("No records found.\n")
  }
}

fetch_nih_data()
